cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp

```diff
@@ -365,17 +444,136 @@ graph_status vertex_stack<Cpu>::increase_stack_size() {
+            if (i != vertex_count_ - 1 && v[i + 1] != null_vertex()) {
+                s.increase_core_level();
+            }
+        }
+        top_ = v;
+    }
+}
+
+template <typename Cpu>
+void global_stack<Cpu>::internal_push(dfs_stack<Cpu>& s, std::uint64_t level) {
+    ONEDAL_ASSERT(vertex_count_ >= 0);
+    // Collect state and push back
+    {
+        const auto v = allocator_.allocate<std::uint64_t>(level + 1);
+
+        for (std::uint64_t i = 0; i < level; ++i) {
+            ONEDAL_ASSERT(i < s.max_level_size);
+            ONEDAL_ASSERT(s.data_by_levels[i].ptop != nullptr);
+            ONEDAL_ASSERT(s.data_by_levels[i].ptop != s.data_by_levels[i].bottom_);
+            ONEDAL_ASSERT(s.data_by_levels[i].ptop >= s.data_by_levels[i].bottom_);
+            ONEDAL_ASSERT(s.data_by_levels[i].ptop <=
+                          s.data_by_levels[i].stack_data + s.data_by_levels[i].stack_size);
+            v[i] = s.data_by_levels[i].ptop[-1];
+        }
+
+        ONEDAL_ASSERT(level < s.max_level_size);
+        ONEDAL_ASSERT(s.data_by_levels[level].ptop != nullptr);
+        ONEDAL_ASSERT(s.data_by_levels[level].bottom_ != nullptr);
+        ONEDAL_ASSERT(s.data_by_levels[level].ptop != s.data_by_levels[level].bottom_);
+        ONEDAL_ASSERT(s.data_by_levels[level].ptop >= s.data_by_levels[level].bottom_);
+        ONEDAL_ASSERT(s.data_by_levels[level].ptop <=
+                      s.data_by_levels[level].stack_data + s.data_by_levels[level].stack_size);
+        v[level] = *(s.data_by_levels[level].bottom_);
+
+        const dal::detail::scoped_lock lock(mutex_);
+        if (size() >= capacity_) {
+            grow();
+        }
+
+        ONEDAL_ASSERT(top_ + vertex_count_ <= bottom_ + capacity_ * vertex_count_);
+        std::uint64_t j = 0;
+        for (; j <= level; ++j) {
+            *(top_++) = v[j];
+        }
+        for (; j < static_cast<std::uint64_t>(vertex_count_); ++j) {
+            *(top_++) = null_vertex();
+        }
+    }
+
+    // Remove state
+    ++(s.data_by_levels[level].bottom_);
+}
+
+template <typename Cpu>
+void global_stack<Cpu>::clear() {
+    if (bottom_ != nullptr) {
+        ONEDAL_ASSERT(top_ != nullptr);
+        allocator_.deallocate(bottom_,
+                              (capacity_ * vertex_count_ > 0) ? capacity_ * vertex_count_ : 1);
+        bottom_ = nullptr;
+        top_ = nullptr;
+        capacity_ = 0;
+    }
+}
+
+template <typename Cpu>
+void global_stack<Cpu>::grow() {
+    const std::int64_t new_capacity = (capacity_ > 0) ? capacity_ * 2 : 1;
+    const auto new_bottom = allocator_.allocate<uint64_t>(
+        (new_capacity * vertex_count_ > 0) ? new_capacity * vertex_count_ : 1);
+    const auto new_top = new_bottom + size() * vertex_count_;
+
+    ONEDAL_IVDEP
+    for (auto dest = new_bottom, src = bottom_; dest != new_top;) {
+        *(dest++) = *(src++);
+    }
+
+    clear();
+
+    bottom_ = new_bottom;
+    top_ = new_top;
+    capacity_ = new_capacity;
```

Comment:
reset ptop?

Reply:
what is ptop? top_ is assigned in line 573.